### ISSUE TITLE
Order of objects on YAML / JSON read / write

### ISSUE THREAD

**Issue by cdeil** [gammapy/gamma-cat#17]
I noticed that after re-running `./make.py all` the order of items in objects kept changing in `docs/data/gammacat-sources.json`, see e.g. this commit: 0858ad46941783ada6e45959380db7634ffdd742

This is bad, because it makes it hard to find real changes in values in the git version history and to review those before committing updates.

I thought this was solved already by making sure `OrderedDict` is always used, see the `gammacat.utils.load_json` helper function (and I think `json.dump` in the `gammacat.utils.write_json` should preserve order).

My guess is that the issue was that we were creating `dict` and lose order on YAML read.
I hope this fixes the issue:
https://github.com/gammapy/gamma-cat/commit/70ed20a7626ce2dd4433c2448e5f4a457632e1e4

I'll keep this issue open as a reminder for a while.
Order in output files should be stable already.
If no problems appear any more this issue can be closed.

**Comment by cdeil** [comment]
This is still an issue ... e.g. you can see the JSON file order jumping around in this commit:
https://github.com/gammapy/gamma-cat/commit/c9263c2b8353cb197832a82132c6e32c42331018

I don't know why.

**Comment by cdeil** [comment]
I've found the issue.
Our current `gammacat.utils.load_yaml` loads the input data as a `dict`, not `OrderedDict`, because we call `yaml.safe_load`.
See http://stackoverflow.com/a/21912744/498873

> I like @James' solution for its simplicity. However, it changes the default global yaml.Loader class, which can lead to troublesome side effects. Especially, when writing library code this is a bad idea. Also, it doesn't directly work with yaml.safe_load().

I'm not checking whether we should switch to the implementation at http://stackoverflow.com/a/21912744/498873 or to https://pypi.python.org/pypi/ruamel.yaml

**Comment by cdeil** [comment]
I switched to `ruamel.yaml` in d5d841b0d4f54d86520e2f84bd3a069bfe55b8f5 .
It's very nice, just has a weird name.
Output files updated (and now in proper order) in 37e2840 .

--

Note to future self: Keys in output JSON files are ordered at the moment because we use `OrderedDict` everywhere, and `json.dumps` then respects the order in JSON serialisation for `OrderedDicts`. For `load_json` we use `json.load(fh, object_pairs_hook=OrderedDict)`, i.e. order is preserved as well.

All is good.
Closing issue.
